client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js

```diff
@@ -1,129 +1,429 @@
-        fallbacks['MultiFactorAuthentication.ADD_FIRST_METHOD']
-      );
+    // Close the modal if the user returns to the introduction screen (ie. they click "back").
+    if (registrationScreen === SCREEN_INTRODUCTION && modalOpen) {
+      this.handleToggleModal();
+    }
+  }
+
+  /**
+   * The backup and default methods are rendered separately
+   * @returns {Array<object>}
+   */
+  getBaseMethods() {
+    const { backupMethod, defaultMethod } = this.props;
+    let { registeredMethods: methods } = this.state;
+
+    if (backupMethod) {
+      methods = methods.filter(method => method.urlSegment !== backupMethod.urlSegment);
+    }
+
+    if (defaultMethod) {
+      methods = methods.filter(method => method.urlSegment !== defaultMethod.urlSegment);
+    }
+
+    return methods;
   }
 
   /**
-   * Handles an event/request to toggle the visibility of the register modal.
+   * Handle a request to toggle the modal. This method will also reset the modal provided it's being
+   * opened, unless `true` is provided as the parameter
+   *
+   * @param {boolean} skipReset
    */
-  handleToggleModal() {
+  handleToggleModal(skipReset = false) {
     const { modalOpen } = this.state;
 
     this.setState({
       modalOpen: !modalOpen,
     });
 
-    if (!modalOpen) {
+    if (!modalOpen && skipReset !== true) {
       // Dispatch a redux action to reset the state of the Register app
       this.props.onResetRegister();
     }
   }
 
   /**
-   * The backup and default methods are rendered separately
-   * @returns {methodShape[]}
+   * Handle a request to add the given method as a registered method
+   *
+   * @param {object} method
    */
-  baseMethods() {
-    const { backupMethod, defaultMethod } = this.props;
-    let { registeredMethods: methods } = this.props;
+  handleAddMethod(method) {
+    this.setState(state => {
+      const { registeredMethods } = state;
 
-    if (backupMethod) {
-      methods = methods.filter(method => method.urlSegment !== backupMethod.urlSegment);
+      // Look for the method if it's existing
+      const existingMethodIndex = registeredMethods.findIndex(
+        candidate => candidate.urlSegment === method.urlSegment
+      );
+
+      if (existingMethodIndex >= 0) {
+        // Replace the existing method
+        registeredMethods[existingMethodIndex] = method;
+        return { registeredMethods };
+      }
+
+      // Otherwise append the new registered method
+      return {
+        registeredMethods: [
+          ...registeredMethods,
+          method,
+        ]
+      };
+    });
+  }
+
+  /**
+   * Handle a request to remove the given method as a registered method
+   *
+   * @param {object} method
+   */
+  handleRemoveMethod(method) {
+    const { onAddAvailableMethod } = this.props;
+
+    this.setState(state => ({
+      registeredMethods: state.registeredMethods.filter(
+        candidate => candidate.urlSegment !== method.urlSegment
+      ),
+    }));
+    onAddAvailableMethod(method);
```

Comment:
There's no guarantee that setState is immediate, you can move this into a callback for the second argument of `setState` to fix this

Reply:
I am using a callback for `setState`? I'm confused.

Reply:
`onAddAvailableMethod` should be the callback right? Its not right now

Reply:
Ok I've refactored this a bit now anyway.